docs/architecture/adr-009-revamp-testkit.md

```diff
@@ -0,0 +1,309 @@
+for encapsulating the possible host variants that need to be covered by `ibc-testkit`.
+
+However, this creates friction in the case when we need to add new host variants. It creates the same
+problem that the `ibc-derive` crate was designed to solve for `ClientState` and
+`ConsensusState` types, namely: dispatching methods to underlying variants of a top-level enum.
+But, a concrete `MockContext` will always have a unique `HostType` and its corresponding `HostBlocks`.
+So we can refactor `HostTypes` and `HockBlocks` with a generic `TestHost` trait that
+maintains its corresponding types e.g. `Block` types, via associated types.
+Finally, we generalize the `MockContext` once more to use this `TestHost` trait
+instead of a concrete enum variant.
+
+This `TestHost` trait should be responsible for generating blocks, headers,
+client, and consensus states specific to that host environment.
+
+```rs
+/// TestHost is a trait that defines the interface for a host blockchain.
+pub trait TestHost: Default + Debug + Sized {
+    /// The type of block produced by the host.
+    type Block: TestBlock;
+
+    /// The type of client state produced by the host.
+    type ClientState: Into<AnyClientState> + Debug;
+
+    /// The type of block parameters to produce a block.
+    type BlockParams: Debug + Default;
+
+    /// The type of light client parameters to produce a light client state.
+    type LightClientParams: Debug + Default;
+
+    /// The history of blocks produced by the host chain.
+    fn history(&self) -> &VecDeque<Self::Block>;
+
+    /// Triggers the advancing of the host chain by extending the history of blocks (or headers).
+    fn advance_block(
+        &mut self,
+        commitment_root: Vec<u8>,
+        block_time: Duration,
+        params: &Self::BlockParams,
+    );
+
+    /// Generate a block at the given height and timestamp, using the provided parameters.
+    fn generate_block(
+        &self,
+        commitment_root: Vec<u8>,
+        height: u64,
+        timestamp: Timestamp,
+        params: &Self::BlockParams,
+    ) -> Self::Block;
+
+    /// Generate a client state using the block at the given height and the provided parameters.
+    fn generate_client_state(
+        &self,
+        latest_height: &Height,
+        params: &Self::LightClientParams,
+    ) -> Self::ClientState;
+}
+
+/// TestBlock is a trait that defines the interface for a block produced by a host blockchain.
+pub trait TestBlock: Clone + Debug {
+    /// The type of header that can be extracted from the block.
+    type Header: TestHeader;
+
+    /// The height of the block.
+    fn height(&self) -> Height;
+
+    /// The timestamp of the block.
+    fn timestamp(&self) -> Timestamp;
+
+    /// Extract the header from the block.
+    fn into_header(self) -> Self::Header;
+}
+
+/// TestHeader is a trait that defines the interface for a header produced by a host blockchain.
```

Comment:
Should this comment be pointing to the "IBC header that submits by relayers" instead of the "header produced by chain"?
(As I see in principle this trait has been implemented on the [TendermintHeader](https://github.com/cosmos/ibc-rs/blob/2acb2fa1324dd838d5165a5819a20cedaa525025/ibc-testkit/src/hosts/tendermint.rs#L200-L216) which if a wrapper of the IBC Header.)

Reply:
Yes, you're right!

Reply:
d9f5c9e